drivers/watchdog/wdt_wwdgt_gd32.c

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2022 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_wwdgt
+
+#include <zephyr/drivers/watchdog.h>
+
+#include <gd32_wwdgt.h>
+#include <soc.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wdt_wwdgt_gd32);
+
+#define WWDGT_NODE DT_NODELABEL(wwdgt)
+
+#define WWDGT_PRESCALER_EXP_MAX (3U)
+#define WWDGT_COUNTER_MIN	(0x40U)
+#define WWDGT_COUNTER_MAX	(0x7fU)
+#define WWDGT_INTERNAL_DIVIDER	(4096ULL)
+
+/* mutable driver data */
+struct gd32_wwdgt_data {
+	/* counter update value*/
+	uint8_t counter;
+	/* user defined callback */
+	wdt_callback_t callback;
+};
+
+static void gd32_wwdgt_irq_config(const struct device *dev);
+
+/**
+ * @param timeout Timeout value in milliseconds.
+ * @param exp exponent part of prescaler
+ *
+ * @return ticks count calculated by this formula.
+ *
+ *  timeout = pclk * INTERNAL_DIVIDER * (2^prescaler_exp) * (count + 1)
+ *  transform as
+ *  count = (timeout * pclk / INTERNAL_DIVIDER * (2^prescaler_exp) ) - 1
+ *
+ *  and add WWDGT_COUNTER_MIN to this as a offset value.
+ */
+static inline uint32_t gd32_wwdgt_calc_ticks(uint32_t timeout, uint32_t exp)
+{
+	const uint32_t pclk = rcu_clock_freq_get(CK_APB1);
+
+	return ((timeout * pclk)
+		/ (WWDGT_INTERNAL_DIVIDER * (1 << exp) * MSEC_PER_SEC) - 1)
+		+ WWDGT_COUNTER_MIN;
+}
```

Comment:
please use clock driver

Reply:
Query the clock frequency with clock_control_get_rate here.